core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -122,6 +122,19 @@ public synchronized void markSeqNoAsCompleted(final long seqNo) {
     }
 
     /**
+     * Resets the checkpoint to the specified value.
+     *
+     * @param checkpoint the local checkpoint to reset this tracker to
+     */
+    synchronized void resetCheckpoint(final long checkpoint) {
+        assert checkpoint <= this.checkpoint;
+        processedSeqNo.clear();
+        firstProcessedSeqNo = checkpoint + 1;
+        nextSeqNo = checkpoint + 1;
```

Comment:
Thinking about this some more, I agree with the assessment we had, except for one thing: We should not reset the `nextSeqNo` variable which is exposed as `getMaxSeqNo`. Otherwise when writing out segments, this max sequence number information which we take from the local checkpoint tracker would be incorrect, i.e. there could be a document in the segment where the sequence number would be above max.

Put differently, nextSeqNo is not tied to the bit set (which represents the pending confirmation marker). Instead it tracks the actual translog.